### ISSUE TITLE
Vulkan + GLFW dynamic binding 

### ISSUE THREAD

**Issue by claudemr** [BindBC/bindbc-glfw#7]
Hello,

I'm making an application using Vulkan and GLFW in D. I use erupted to load Vulkan symbols (it works fine with Bindbc.SDL).
It looks like bindGLFW_Vulkan (defined at binddynamic.d:476) does not compile as it is supposed to when used with mixin, as described in the comment just above line 456 (which looks inaccurate, symbol glfwVulkanBind does not exist, I suppose it was meant to be bindGLFW_Vulkan).

Here's my code:
```
module myapp;
import erupted;  // Necessary to have VK symbols object, used in bindGLFW_Vulkan
import bindbc.glfw;

mixin(bindGLFW_Vulkan); //Compilation fails: lib, errorCount and invalidHandle are undefined
```

The reason it fails is because the 3 undefined symbols are private within binddynamic.d.

To solve the problem, I defined some public functions hub to lib and errorCount in binddynamic.d:
```
public auto getGLFWlib()
{
    return lib;
}
    
public auto getGLFWerrorCount()
{
    return errorCount();
}
```

And I changed the definition of loadGLFW_Vulkan to:
```
bool loadGLFW_Vulkan()
{
    import bindbc.loader;
    assert(isGLFWLoaded(), "loadGLFW must be successfully called before loadGLFW_Vulkan");

    auto errCount = getGLFWerrorCount();
    getGLFWlib.bindSymbol(cast(void**)&glfwGetRequiredInstanceExtensions, "glfwGetRequiredInstanceExtensions");
    getGLFWlib.bindSymbol(cast(void**)&glfwGetInstanceProcAddress, "glfwGetInstanceProcAddress");
    getGLFWlib.bindSymbol(cast(void**)&glfwGetPhysicalDevicePresentationSupport, "glfwGetPhysicalDevicePresentationSupport");
    getGLFWlib.bindSymbol(cast(void**)&glfwCreateWindowSurface, "glfwCreateWindowSurface");
    return getGLFWerrorCount() == errCount;
}
```

With those changes, it builds fine, and it works.

I suspect the same problems occurs with bindGLFW_EGL, bindGLFW_WGL, bindGLFW_Windows, bindGLFW_NSGL, bindGLFW_Cocoa (enum name mismatches at line 595), bindGLFW_GLX, bindGLFW_X11, etc...

I have not made a pull-request for that as it requires some API change, and the names  getGLFWlib and getGLFWerrorCount I used are pretty bad. Maybe there is a more elegant solution (or maybe I have missed something)?

If you need me to do the mundane task of writing the correction, I can make a pull-request for you to review as the code is fresh in my mind.

Regards,

Claude

**Comment by mdparker** [comment]
Thanks for reporting this. That I never noticed this, and that no one has reported it, shows how often these optional functions are used. I'm considering different ways to handle it. I'm leaning toward making it configurable via Versions on the command line, like I've handled OpenGL extensions. Then the functions can be automatically declared directly in the same module as the other GLFW functions and loaded automatically.

**Comment by claudemr** [comment]
Using Versions sounds like a good idea: that would also make the API more consistent.

I am actually following the tutorials from https://vulkan-tutorial.com which use GLFW. I was surprised to be the first to bump into that issue at first. I think it's due to the combination Vulkan + D. I think those optional functions will be used more and more as Vulkan and D both increase their adoption.

**Comment by mdparker** [comment]
I've kept this on the backburner for higher priority stuff. I finally made time to implement it today, but then I remembered why I had used mixins in the first place -- some of this stuff requires platform-specific declarations that need to come from different bindings, e.g., EGL and Vulkan. I can't have dependencies on specific EGL or Vulkan bindings. And if I declare my own local types just to get things to compile, that could cause compatibility issues. The mixins allow users to import whatever binding they want and, as long as it uses the same declarations as the C headers, mixin the GLFW stuff in the same module.

I'll have to look into this more deeply. I believe the relevant types are all just pointers, so I may be able to get away with just using `void*`. If that's the case, I can go ahead and implement things to use `version`. Otherwise, I'll have to figure out another approach to keep the mixins.

**Comment by claudemr** [comment]
Hello,

Overall, I think the "mixin" approach is better compared to using "void*" or generic data-types to make everything compile. It feels more "correct" to me, and probably more future-proof. Even though it looks a bit awkward.

I don't have a strong opinion about this, it's quite a design decision.

The solution I'm using works fine (having a public proxy to GLFW_lib handle and errorCount()), but it's not very elegant... :)

Claude

**Comment by mdparker** [comment]
I don't know what took me so long to get to the easy answer: a new public function in `dynload.d` called `loadGLFWSymbol` that forwards to `lib.bindSymbol`, and private imports of `bindbc.loader : errorCount` in the `loadGLFW_*` functions . All available in 0.10.0, along with documentation in the README. Thanks again for reporting this!

**Comment by claudemr** [comment]
Hello, sorry for the late reply, it works fine with Vulkan. Thanks.

I'll submit a PR to fix a minor typo problem in loadGLFW_Vulkan.
